### PR TITLE
v4.0.0 Added docker support

### PR DESCRIPTION
**Pull request by antonplagemann**
New features:
- Added full docker 🐳 support with automatic deployments

Breaking changes:
- Changed config from python file (`conf.py`) to `.env` file (for update instructions see below)

Bugfixes:
- Fixed missing gender error #29
- Fixed an issue where a raw birthday format produced an exception instead of beeing skipped
- Fixed an issue where some contacts have been unintentionally skipped during delta sync
- Fixed an issue where a new Google token would not have a `refresh_token` for non-interactive token refresh
- Fixed an issue where OAuth scopes have been parsed wrong during Google token refresh
- Fixed an issue where a deleted contact does not get removed from the database (=> failing database check)
- Fixed a rare issue where database check could fail
- Fixed and improved some logging issues

Background changes (no effects on functionality):
- Added instructions for setting up a GCP People API project
- Added fully automated tests of every command and environment using GitHub Actions
- Added SonarCloud SAST CI
- Added CodeQL SAST CI
- Added a new command-line switch for specifying config location path
- Introduced `black` codestyle
- Introduced `flake8` linting
- Introduced `bandit` security check
- Introduced `mypy` type checking
- Introduced `isort` import sorting
- Reduced code complexity and duplication
- Introduced snake_case naming convention
- Added custom exceptions (better error handling)
- Changed initial Google authorization flow from `local_webserver` to `console` (more robust)
- Changed pickling format from `binary` to `base64` to be readable characters (file updates automatically on next script run)

**Update instructions from v3.x.x**:
- **Automatic**: Do a `pip install -r requirements.txt`, then run `GMSync.py -u` once. This converts your `conf.py` into a `.env` file and moves your credential files into the right subfolders.
- Manual: Create a new `.env` file (use `.env.example` as template. Then move your `credentials.json` and `token.pickle` files into a new `data` subfolder. Do a `pip install -r requirements.txt` afterwards.

### CODE REVIEW

File: test/docker-compose-sync.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+      - ../logs:/usr/app/logs
+
+    # Adjust command if needed (-u needed for getting console output)
+    command: python -u GMSync.py --help
+    networks:
+      - monica_network
+
+networks:
+  monica_network:
+    name: monica_network
```
Comment (p-fruck) [inline test/docker-compose-sync.yml:28]: The newline!

**Review by p-fruck — CHANGES_REQUESTED** [review]

File: .github/workflows/python-ci.yml [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+      - name: Prepare initial sync
+        run: python test/ChaosMonkey.py --initial --num ${{  github.event.inputs.numChaos || 4  }}
+      - name: Run initial sync
+        run: docker-compose -f test/docker-compose-sync.yml -f test/docker-compose-sync-initial.yml --env-file .env up
+        timeout-minutes: 5
+      - name: Check initial sync results
+        if: always()
+        run: |
+          mv logs/sync.log logs/${{ env.TEST_RUNNER }}_sync_initial.log && \
+          [[ -z "$(grep -e ERROR -e WARNING logs/${{ env.TEST_RUNNER }}_sync_initial.log)" ]] || exit 1
```
[the remark this answers is not in the document]
Comment (antonplagemann) [reply .github/workflows/python-ci.yml:276]: Hm, I will ask the expert that created the expression for me 🤔 
Wait... that was you... Ok, I'll remove it 😆

File: docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+version: "3.8"
+
+services:
+  python:
+    image: antonplagemann/google-monica-sync
```
[the remark this answers is not in the document]
Comment (antonplagemann) [reply docker-compose.yml:5]: Ok, it's implicit the `latest` tag but it's maybe better to specify it explicitly 👍

File: .vscode/settings.json [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+        "--no-pretty",
+        "--install-types"
+    ],
+    "python.linting.mypyEnabled": true,
+    "cSpell.words": [
+        "dotenv"
+    ],
+    //"python.envFile": "${workspaceFolder}/.env",
+    
+}
```
[the remark this answers is not in the document]
Comment (antonplagemann) [reply .vscode/settings.json:34]: Comments are allowed in vscode settings JSON files 😆

**Comment by antonplagemann** [comment]
@p-fruck 
Many thanks for doing a review! 🙏 ❤
Much appreciated! 🚀

File: .github/actions/setup-environment/action.yml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    - name: Create API token at Monica instance
+      run: python test/SetupToken.py
+      shell: bash
+
+    - name: Create database file and monkey state
+      run: python test/ChaosMonkey.py
+      shell: bash
+
+    - name: Set folder permissions for non-root containers
+      run: sudo chmod 777 data logs -R
```
[the remark this answers is not in the document]
Comment (antonplagemann) [reply .github/actions/setup-environment/action.yml:63]: I don't know exactly why, but it seems required as the pipeline is now failing with a permission error during script execution inside the container 🤷‍♂️

File: .github/workflows/python-ci.yml [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+      # Test delta sync
+      - name: Prepare delta sync
+        run: python test/ChaosMonkey.py --delta --num ${{  github.event.inputs.numChaos || 4  }}
+      - name: Run delta sync
+        run: docker-compose -f test/docker-compose-sync.yml -f test/docker-compose-sync-delta.yml --env-file .env up
```
[the remark this answers is not in the document]
Comment (antonplagemann) [reply .github/workflows/python-ci.yml:282]: Actually this is required here, thats a bit tricky.
As the compose file is in a subfolder, `docker-compose` will only look inside the subfolder for the `.env` file. To override this behaviour, I need to specify the `.env` path relatively to the `cwd`, which, in this case, is just `.env` 😆

**Review by p-fruck — APPROVED** [review]

**Pull request merged**